core/base/integer_range.hpp

```diff
@@ -0,0 +1,268 @@
+    iterator end_;
+};
+
+
+template <typename IndexType>
+class irange_strided : public random_access_range<integer_iterator<IndexType>> {
+    using base = random_access_range<integer_iterator<IndexType>>;
+
+public:
+    using index_type = IndexType;
+    using iterator = typename base::iterator;
+    using value_type = typename base::value_type;
+    using reference = typename base::reference;
+    using difference_type = typename base::difference_type;
+
+    explicit irange_strided(index_type begin, index_type end, index_type stride)
+        : random_access_range<integer_iterator<IndexType>>{
+              iterator{begin, stride},
+              iterator{begin + (end - begin) / stride * stride +
+                           ((end - begin) % stride != 0 ? stride : 0),
+                       stride}}
+    {
+        assert(end >= begin);
+        assert(stride > 0);
+    }
+
+    constexpr friend bool operator==(irange_strided lhs, irange_strided rhs)
+    {
+        return lhs.stride() == rhs.stride() && *lhs.begin() == *rhs.begin() &&
+               *lhs.end() == *rhs.end();
+    }
+
+    constexpr friend bool operator!=(irange_strided lhs, irange_strided rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    constexpr index_type begin_index() const { return *this->begin(); }
+
+    constexpr index_type end_index() const { return *this->end(); }
+
+    constexpr index_type stride() const { return this->begin().stride(); }
+};
+
+
+template <typename IndexType>
+class irange : public random_access_range<integer_iterator<IndexType>> {
+    using base = random_access_range<integer_iterator<IndexType>>;
+
+public:
+    using index_type = IndexType;
+    using iterator = typename base::iterator;
+    using value_type = typename base::value_type;
+    using reference = typename base::reference;
+    using difference_type = typename base::difference_type;
+
+    explicit irange(index_type begin, index_type end)
+        : random_access_range<integer_iterator<IndexType>>{iterator{begin, 1},
+                                                           iterator{end, 1}}
+    {
+        assert(end >= begin);
+    }
+
+    explicit irange(index_type end) : irange{0, end} {}
+
+    constexpr index_type begin_index() const { return *this->begin(); }
+
+    constexpr index_type end_index() const { return *this->end(); }
+
+    template <typename Group>
+    constexpr irange_strided<index_type> striped(Group g) const
+    {
+        return striped(detail::group_traits<Group>::get_local_id(g),
+                       detail::group_traits<Group>::get_size(g));
+    }
+
+    constexpr irange_strided<index_type> striped(index_type local_index,
+                                                 index_type group_size) const
+    {
+        assert(local_index >= 0);
+        assert(local_index < group_size);
+        return irange_strided<index_type>{begin_index() + local_index,
+                                          end_index(), group_size};
+    }
```

Comment:
They are targeting SIMD/coalescing memory accesses in general. My aim is to turn the first loop into the second:

```cpp
for (auto i = begin + warp.thread_rank(); i < end; i += warp.size());
for (auto i : irange{begin, end}.striped(warp))
```

I will remove the templated version though, since that will be part of a separate PR.